File: src/daily_report/collectors/github_trending.py
from __future__ import annotations

import re
from datetime import date
from typing import Any
import httpx

TRENDING_URL = "https://github.com/trending?since=daily"


def fetch_trending_html() -> str:
    headers = {
        "User-Agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        "Accept-Language": "en-US,en;q=0.9",
    }
    with httpx.Client(timeout=20, follow_redirects=True, headers=headers) as client:
        resp = client.get(TRENDING_URL)
        resp.raise_for_status()
        return resp.text


def parse_trending(html: str) -> list[dict[str, Any]]:
    articles = re.findall(r'<article class="Box-row">(.*?)</article>', html, re.DOTALL)
    items: list[dict[str, Any]] = []
    for rank, block in enumerate(articles, start=1):
        name_m = re.search(r'<h2[^>]*>\s*<a[^>]+href="/([^"]+)"', block)
        desc_m = re.search(r'<p[^>]*class="col-9 color-fg-muted my-1 pr-4">(.*?)</p>', block, re.DOTALL)
        lang_m = re.search(r'programmingLanguage">(.*?)</span>', block, re.DOTALL)
        stars_m = re.search(r'href="/[^"]+/stargazers"[^>]*>\s*([\d,]+)\s*</a>', block)
        today_m = re.search(r'([\d,]+)\s+stars today', block)
        if not name_m:
            continue
        desc = re.sub(r'<.*?>', '', desc_m.group(1)).strip() if desc_m else ''
        lang = re.sub(r'<.*?>', '', lang_m.group(1)).strip() if lang_m else ''
        items.append({
            "rank": rank,
            "repo": name_m.group(1).strip(),
            "description": re.sub(r'\s+', ' ', desc),
            "language": lang,
            "stars": int((stars_m.group(1) if stars_m else '0').replace(',', '')),
            "stars_today": int((today_m.group(1) if today_m else '0').replace(',', '')),
            "url": f"https://github.com/{name_m.group(1).strip()}",
        })
    return items


def collect() -> dict[str, Any]:
    html = fetch_trending_html()
    items = parse_trending(html)
    return {
        "source": "github_trending",
        "collected_at": date.today().isoformat(),
        "count": len(items),
        "items": items,
    }

File: src/daily_report_js/collectors/github_trending.js
export async function fetchGithubPage(since = 'daily', lang = '') {
  let url = 'https://github.com/trending';
  if (lang) url += `/${lang}`;
  url += `?since=${since}`;

  const resp = await fetch(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      'Accept-Language': 'en-US,en;q=0.9',
    },
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
  return await resp.text();
}

async function fetchReadmeExcerpt(repoUrl) {
  const resp = await fetch(repoUrl, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
      'Accept-Language': 'en-US,en;q=0.9',
    },
  });
  if (!resp.ok) throw new Error(`README HTTP ${resp.status}`);

  const html = await resp.text();
  const article = html.match(/<article[^>]*class="[^"]*markdown-body[^"]*"[^>]*>([\s\S]*?)<\/article>/)?.[1] || '';
  if (!article) return '';

  const paragraphs = [...article.matchAll(/<p[^>]*>([\s\S]*?)<\/p>/g)]
    .map((m) => m[1])
    .map((s) => s
      .replace(/<br\s*\/?>/gi, ' ')
      .replace(/<[^>]+>/g, ' ')
      .replace(/&nbsp;/g, ' ')
      .replace(/&amp;/g, '&')
      .replace(/&lt;/g, '<')
      .replace(/&gt;/g, '>')
      .replace(/&quot;/g, '"')
      .replace(/&#39;/g, "'")
      .replace(/&#\d+;/g, ' ')
      .replace(/\s+/g, ' ')
      .trim())
    .filter(Boolean)
    .filter((s) => s.length >= 20)
    .slice(0, 2);

  return paragraphs.join('\n\n').trim();
}

async function enrichMissingDescriptions(items) {
  for (const item of items) {
    if (String(item.description || '').trim()) continue;
    try {
      const excerpt = await fetchReadmeExcerpt(item.url);
      if (excerpt) {
        item.description = excerpt;
        console.log(`  ↪ README fallback: ${item.repo}`);
      }
    } catch (err) {
      console.log(`  ⚠ README fallback failed for ${item.repo}: ${err.message || String(err)}`);
    }
  }
}

export function parseGithubTrending(html) {
  const articles = [...html.matchAll(/<article class="Box-row">([\s\S]*?)<\/article>/g)].map((m) => m[1]);
  const items = [];

  articles.forEach((block, i) => {
    const name = block.match(/<h2[^>]*>\s*<a[^>]+href="\/([^"]+)"/)?.[1]?.trim() || '';
    if (!name) return;

    const descRaw = block.match(/<p[^>]*class="[^"]*col-9[^"]*"[^>]*>([\s\S]*?)<\/p>/)?.[1] || '';
    const desc = descRaw.replace(/<[^>]+>/g, '').replace(/\s+/g, ' ').trim();

    const lang = block.match(/itemprop="programmingLanguage"[^>]*>(.*?)<\/span>/)?.[1]?.trim() || '';
    const starsTotal = Number((block.match(/href="\/[^/]+\/[^/]+\/stargazers"[^>]*>.*?<\/svg>\s*([\d,]+)/s)?.[1] || '0').replaceAll(',', ''));
    const forks = Number((block.match(/href="\/[^/]+\/[^/]+\/forks"[^>]*>.*?<\/svg>\s*([\d,]+)/s)?.[1] || '0').replaceAll(',', ''));
    const starsToday = Number((block.match(/([\d,]+)\s*stars today/)?.[1] || '0').replaceAll(',', ''));

    items.push({
      rank: i + 1,
      repo: name,
      url: `https://github.com/${name}`,
      description: desc,
      language: lang,
      stars_total: starsTotal,
      forks,
      stars_today: starsToday,
    });
  });

  return items;
}

export async function collectGithub({ since = 'daily', lang = '' }) {
  const html = await fetchGithubPage(since, lang);
  const items = parseGithubTrending(html);
  await enrichMissingDescriptions(items);
  const now = new Date();

  return {
    source: 'github_trending',
    collected_at: now.toISOString(),
    since,
    lang_filter: lang || 'all',
    count: items.length,
    items,
  };
}
